### PR TITLE
Bpk 2420 implement star rating on android

### PR DESCRIPTION
**Pull request by bvitaliyg**
+ [ ] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/master/CONTRIBUTING.md)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/master/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+  @Test
+  fun screenshotTestStarRating_OverMaxValue() {
+    rating.maxRating = 5
+    rating.rating = 6.0f
+    snap(rating)
+  }
+
+  @Test
+  fun screenshotTestStarRating_StarColor() {
```
Comment (tiagoengel) [inline app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt]: Can we remove both of this tests and add a `withTheme` test. Have a look at this to see an example: https://github.com/Skyscanner/backpack-android/blob/master/app/src/androidTest/java/net/skyscanner/backpack/toggle/BpkSwitchTest.kt#L31

File: app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+  @Test
+  fun screenshotTestStarRating_CustomMaxRatingDecreasing() {
+    rating.maxRating = 5
+    rating.maxRating = 3
+    snap(rating)
+  }
+
+  @Test
+  fun screenshotTestStarRating_NegativeValue() {
```
Comment (tiagoengel) [inline app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt]: Can we change to `NegativeRating`?

File: app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    rating.maxRating = 5
+    rating.rating = -0.5f
+    snap(rating)
+  }
+
+  @Test
+  fun screenshotTestStarRating_ZeroValue() {
```
Comment (tiagoengel) [inline app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt]: Same here and for all the rating tests

File: app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    rating.maxRating = 5
+    rating.rating = 0.0f
+    snap(rating)
+  }
+
+  @Test
+  fun screenshotTestStarRating_BeforeHalfStarValue() {
```
Comment (tiagoengel) [inline app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt]: Can we change the name of those tests? I'm not sure what Before/OverHalf means here. We can name it something like `RatingValueBetween0And0_5` for example

File: app/src/main/java/net/skyscanner/backpack/demo/data/ComponentRegistry.kt [hunk 5]
```diff
@@ -123,6 +123,12 @@ object ComponentRegistry {
         "Emphasized" story NodeData { Story of R.layout.fragment_text_emphasized },
         "Heavy" story NodeData { Story of R.layout.fragment_text_heavy },
         "With drawables" story NodeData { Story of R.layout.fragment_text_drawables }
+      )),
+    "Star Rating" story NodeData({ children -> SubStory of children },
+      mapOf(
+        "Default" story NodeData { Story of R.layout.fragment_star_rating_default },
+        "Different values" story NodeData { Story of R.layout.fragment_star_rating_values },
+        "Custom Max Rating" story NodeData { Story of R.layout.fragment_star_rating_max }
```
Comment (tiagoengel) [inline app/src/main/java/net/skyscanner/backpack/demo/data/ComponentRegistry.kt]: Can we add one to test RTL? it's easier to have a quick look when we need it

File: app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+import android.graphics.Color
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import net.skyscanner.backpack.BpkSnapshotTest
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+
+class BpkStarRatingTest : BpkSnapshotTest() {
```
Comment (tiagoengel) [inline app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt]: Can we also add a test with RTL?

File: app/src/main/res/values/themes.xml [hunk 7]
```diff
@@ -39,6 +40,11 @@
     <item name="fontFamilyEmphasized">@font/shadows_into_light</item>
   </style>
 
+  <style name="LondonThemeBpkStarRating" parent="">
+    <item name="starColor">?bpkGray100Color</item>
+    <item name="starFilledColor">?colorAccent</item>
```
Comment (tiagoengel) [inline app/src/main/res/values/themes.xml]: Can we change this to `?npkPrimaryColor`? We are not setting `colorAccent` on those  themes

File: Backpack/src/main/java/net/skyscanner/backpack/starrating/BpkStarRating.kt [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      }
+    }
+
+  init {
+    initialize(attrs, defStyleAttr)
+  }
+
+  private fun initialize(attrs: AttributeSet?, defStyleAttr: Int) {
+    orientation = HORIZONTAL
+    attrs
```
Comment (tiagoengel) [inline Backpack/src/main/java/net/skyscanner/backpack/starrating/BpkStarRating.kt]: This code should run if attrs is null so the themed props are fetched. Don't need to check for null `obtainStyledAttributes` supports passing null as argument

File: app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt (hunk 1 above)
Comment (bvitaliyg) [reply app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt]: Done

File: app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt (hunk 2 above)
Comment (bvitaliyg) [reply app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt]: done

File: app/src/main/java/net/skyscanner/backpack/demo/data/ComponentRegistry.kt (hunk 5 above)
Comment (bvitaliyg) [reply app/src/main/java/net/skyscanner/backpack/demo/data/ComponentRegistry.kt]: Done

File: app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt (hunk 4 above)
Comment (bvitaliyg) [reply app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt]: Done

File: app/src/main/res/values/themes.xml (hunk 7 above)
Comment (bvitaliyg) [reply app/src/main/res/values/themes.xml]: Done

File: Backpack/src/main/java/net/skyscanner/backpack/starrating/BpkStarRating.kt (hunk 8 above)
Comment (bvitaliyg) [reply Backpack/src/main/java/net/skyscanner/backpack/starrating/BpkStarRating.kt]: Done

File: app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt (hunk 3 above)
Comment (bvitaliyg) [reply app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt]: Done

File: app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt (hunk 6 above)
Comment (bvitaliyg) [reply app/src/androidTest/java/net/skyscanner/backpack/starrating/BpkStarRatingTest.kt]: Done

File: Backpack/src/main/java/net/skyscanner/backpack/starrating/BpkStarRating.kt [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        value = value,
+        starFilledColor = starFilledColor,
+        starColor = starColor
+      )
+    }
+  }
+}
+
+private fun wrapContext(context: Context, attrs: AttributeSet?): Context {
+  val withBaseStyle = wrapContextWithDefaults(context)
```
[the remark this answers is not in the document]
Comment (bvitaliyg) [reply Backpack/src/main/java/net/skyscanner/backpack/starrating/BpkStarRating.kt]: Done

**Comment by georgegillams** [comment]
Thank you for your contribution.
You can use the following dependency to test your application while we review your pull request:

`com.github.Skyscanner:backpack-android:269646eefb746b81e3578c0b5f529ef9e7425585`

**Pull request merged**